lib/bots/keyframebot.ts

```diff
@@ -0,0 +1,683 @@
+                    `Base64 encoded! Aborting.`);
+                throw new HTTPError(500, 'Keyframe was not correctly encoded');
+            }
+
+            // We now go ahead and:
+            // 1. Create a new branch for this, create it from the version passed
+            // 2. Commit the keyframe to that branch
+            // 3. Open a new PR pointing to that branch. Any relevant reviewers can be set (when it works) from
+            //    a `.procbot.yml` config in the env repo.
+
+            // Create a new branch and commit the keyframe to it.
+            deployDetails = {
+                keyframe: keyframeFile,
+                username: decodedToken.username,
+                environment,
+                version,
+                owner,
+                repo
+            };
+            return this.createNewEnvironmentBranchCommit(deployDetails);
+        }).then((branchName: string) => {
+            // Open a new PR using the new branch.
+            // If there's a `.procbot.yml` config in the branch, it'll do setup for us.
+            return this.dispatchToEmitter(this.githubEmitterName, {
+                data: {
+                    owner,
+                    repo,
+                    title: `Merge product keyframe ${deployDetails.version} into ${deployDetails.environment}`,
+                    body: `PR was created via a deployment of the keyframe by Resin admin ${deployDetails.username}.`,
+                    head: branchName,
+                    base: 'master'
+                },
+                method: this.githubApi.pullRequests.create
+            });
+        }).then(() => {
+            // Badabing. We'll now do linting on the *environment* branch automatically, as the
+            // PR will kick it off. NOTE: How do we determine which type of linting we do?
+            // I guess we could look at repo, but that's a bit horrible. See if there's a variables file?
+            // Talk to Jack.
+            res.sendStatus(200);
+        }).catch((err: GithubError | HTTPError) => {
+            let errorCode = (err instanceof HTTPError) ? err.httpCode : 500;
+            this.reportError(err);
+            res.status(errorCode).send(err.message);
+        });
+    }
+
+    /**
+     * Creates a new branch on the specified environment for a given keyframe.
+     * @param branchDetails The details of the keyframe, user and version.
+     * @returns             A string naming the branch created on fulfilment.
+     */
+    private createNewEnvironmentBranchCommit = (branchDetails: DeploymentDetails): Promise<string> => {
+        const owner = branchDetails.owner;
+        const repo = branchDetails.repo;
+        const keyframe = branchDetails.keyframe;
+        const environment = branchDetails.environment;
+        const version = branchDetails.version;
+        const user = branchDetails.username;
+        const branchName = `${user}-${version}`;;
```

Comment:
Should the `branchName` include the `${environment}` too? Otherwise an error might get thrown if the user tries pushing the same version to two different environments? (or perhaps I'm just misunderstanding this)

EDIT: Ahhh, I think I now see that there's a 1:1 mapping of environments to repos (i.e. each environment is hosted by separate repos). So should KeyframeBot at startup check that two separate environments don't map to the same repo?